trace_decoder/src/decoding.rs

```diff
@@ -219,542 +41,510 @@ struct TrieDeltaApplicationOutput {
     additional_storage_trie_paths_to_not_hash: HashMap<H256, Vec<Nibbles>>,
 }
 
-impl ProcessedBlockTrace {
-    pub fn into_txn_proof_gen_ir(
-        self,
-        other_data: OtherBlockData,
-    ) -> TraceParsingResult<Vec<GenerationInputs>> {
-        let mut curr_block_tries = PartialTrieState {
-            state: self.tries.state.as_hashed_partial_trie().clone(),
-            storage: self
-                .tries
-                .storage
-                .iter()
-                .map(|(k, v)| (*k, v.as_hashed_partial_trie().clone()))
-                .collect(),
-            ..Default::default()
-        };
-
-        let mut extra_data = ExtraBlockData {
-            checkpoint_state_trie_root: other_data.checkpoint_state_trie_root,
-            txn_number_before: U256::zero(),
-            txn_number_after: U256::zero(),
-            gas_used_before: U256::zero(),
-            gas_used_after: U256::zero(),
-        };
+pub fn into_txn_proof_gen_ir(
+    ProcessedBlockTrace {
+        tries: PartialTriePreImages { state, storage },
+        txn_info,
+        withdrawals,
+    }: ProcessedBlockTrace,
+    other_data: OtherBlockData,
+) -> anyhow::Result<Vec<GenerationInputs>> {
```

Comment:
In its current form calling this a "library" would probably be a stretch of the imagination to some extent. It's neither library nor application right now, sure the intention was (and is) that it should be a library but it's not really today.

Reply:
I think you're right to be skeptical.

I'm also far from an expert on these crates, but the error paths are typically bubbled straight to the user.
I like `thiserror` for when the user needs to be able to _handle_ error cases:
```rust
if let Err(e) = foo() {
    match e {
        ...
    }
}
```
Which currently *never happens* (as you alluded to), and we're paying a substantial cost for no benefit (c.f line count)

> The question is whether it would be more appropriate to change the public functions to expose fine-grained errors rather than anyhow. Changing the internal code to also use anyhow will make it harder to do in the future if the need arises.

If this crate remains standing [after the dust is settled](#275), I'd be sorely tempted to use a more careful error design that hides causes altogether, c.f [`serde_json::Error::classify`](https://docs.rs/serde_json/latest/serde_json/struct.Error.html#method.classify) or [`request::Error`](https://docs.rs/reqwest/latest/reqwest/struct.Error.html), or even the grandaddy of them all, [`io::Error`](https://matklad.github.io/2020/10/15/study-of-std-io-error.html).

But for now I hope you'll agree that the complexity tradeoff is worth it, especially in the face of refactoring to come.

Reply:
@iljakuklic do you reckon this is a blocker?

Reply:
> @iljakuklic do you reckon this is a blocker?

From my side it's not fwiw

Reply:
> But for now I hope you'll agree that the complexity tradeoff is worth it, especially in the face of refactoring to come.

I don't think I have enough context to make the determination at the moment.

> @iljakuklic do you reckon this is a blocker?

No, just something to consider. I will leave the decision to your judgement.